### PR TITLE
maybe working shortest path

### PR DESCRIPTION
**Pull request by kawa2287**
- Updated the old recursion function.  Should now use the `Match` class similar to the iterative function.
- Updated old tests and added a new one

### CODE REVIEW

File: title_belt_nhl/schedule.py [hunk 1]
```diff
@@ -139,23 +140,32 @@ def find_match(self, current_belt_holder, from_date) -> Match:
...
                 return match
 
-    def find_nearest_path_str(self, teams, path_string, from_date=None) -> str:
-        newTeams = []
-        if from_date:
-            self.set_from_date(from_date)
-        for tm in teams:
-            splits = tm.split(" -> ")
-            cur_match: Match = self.find_match(splits[-1], self.from_date)
+    def find_nearest_path_v2(self, scenarios: list[list[Match]]) -> list[Match]:
```
Comment (elffaW) [inline title_belt_nhl/schedule.py]: Passing in a `[[Match]]` feels super strange, but maybe it's necessary? Recursion is annoying.

File: title_belt_nhl/schedule.py [hunk 2]
```diff
@@ -139,23 +140,32 @@ def find_match(self, current_belt_holder, from_date) -> Match:
...
+                    self.createNewScenarioBranch(cur_match.home, cur_match.serial_date, s)
+                )
+                newScenarios.append(
+                    self.createNewScenarioBranch(cur_match.away, cur_match.serial_date, s)
+                )
+
+        shortestPath = self.find_nearest_path_v2(newScenarios)
+        return shortestPath
+
+    def createNewScenarioBranch(self, team: str, matchDate: int, scenario: list[Match]):
```
Comment (elffaW) [inline title_belt_nhl/schedule.py]: `snake_case` 🐍 🐍 🐍

File: title_belt_nhl/schedule.py [hunk 3]
```diff
@@ -139,23 +140,32 @@ def find_match(self, current_belt_holder, from_date) -> Match:
...
+        scenario_copy = copy.deepcopy(scenario)
```
Comment (elffaW) [inline title_belt_nhl/schedule.py]: I don't remember if it's more efficient or not, but if you're only using `deepcopy` then consider just importing only that: `from copy import deepcopy`.

File: title_belt_nhl/tests/test_schedule.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
 
 MOCK_DATA_PATH = "./title_belt_nhl/tests/test_files/mock_league_schedule.json"
 MOCK_DATA_PATH_BIG = "./title_belt_nhl/tests/test_files/mock_league_schedule_big.json"
+MOCK_DATA_PATH_ALL = "./title_belt_nhl/tests/test_files/mock_league_schedule_all.json"
```
Comment (elffaW) [inline title_belt_nhl/tests/test_schedule.py]: Do we need the entire schedule or should we just add a few days to the `big` one?

File: title_belt_nhl/schedule.py [hunk 5]
```diff
@@ -139,23 +140,32 @@ def find_match(self, current_belt_holder, from_date) -> Match:
...
+                # Add new Scenario branches
+                newScenarios.append(
+                    self.createNewScenarioBranch(cur_match.home, cur_match.serial_date, s)
+                )
+                newScenarios.append(
+                    self.createNewScenarioBranch(cur_match.away, cur_match.serial_date, s)
+                )
+
+        shortestPath = self.find_nearest_path_v2(newScenarios)
+        return shortestPath
```
Comment (elffaW) [inline title_belt_nhl/schedule.py:183]: What happens if the season ends and no match is found?

File: title_belt_nhl/tests/test_schedule.py [hunk 6]
```diff
@@ -134,3 +159,34 @@ def test_find_nearest_path_games_big(self, monkeypatch, league_schedule_big):
...
+
+        m1 = Match("FLA", "BOS", date_obj=date(2024, 10, 8))
+        m2 = Match("OTT", "FLA", date_obj=date(2024, 10, 10))
+        m3 = Match("BUF", "FLA", date_obj=date(2024, 10, 12))
+        m4 = Match("PIT", "BUF", date_obj=date(2024, 10, 16))
+        m5 = Match("CBJ", "BUF", date_obj=date(2024, 10, 17))
+        m6 = Match("CBJ", "MIN", date_obj=date(2024, 10, 19))
+        expected = [m1, m2, m3, m4, m5, m6]
+        for i, m in enumerate(expected):
+            assert str(path_matches[i]) == str(m)
```
Comment (elffaW) [inline title_belt_nhl/tests/test_schedule.py:170]: TODO: also check `belt_holder` property is as-expected for all of these tests.

**Review by elffaW — APPROVED** [review]
Seems like an improvement, and found a shorter path for `MIN` than the iterative version!

**Review by elffaW — CHANGES_REQUESTED** [review]
Oh whoops.

- fix main.py
- maybe handle no-match scenario?

File: title_belt_nhl/schedule.py (hunk 2 above)
Comment (elffaW) [reply title_belt_nhl/schedule.py]: 👍 🐍 
👎 🐫

File: title_belt_nhl/schedule.py (hunk 5 above)
Comment (kawa2287) [reply title_belt_nhl/schedule.py:183]: We get paid.

File: title_belt_nhl/schedule.py (hunk 1 above)
Comment (kawa2287) [reply title_belt_nhl/schedule.py]: If you have other ideas/ suggestions... I'm open to it haha

File: title_belt_nhl/schedule.py (hunk 5 above)
Comment (kawa2287) [reply title_belt_nhl/schedule.py:183]: Attempted to handle this [here](https://github.com/kawa2287/title-belt-nhl/blob/d4fc48bc18d55d9daa3f7bba275dd2b9f367891f/title_belt_nhl/schedule.py#L154C1-L160C24).   And added tests [here](https://github.com/kawa2287/title-belt-nhl/blob/d4fc48bc18d55d9daa3f7bba275dd2b9f367891f/title_belt_nhl/tests/test_schedule.py#L173C1-L199C36)

File: title_belt_nhl/tests/test_schedule.py (hunk 6 above)
Comment (kawa2287) [reply title_belt_nhl/tests/test_schedule.py:170]: Added this test for `test_find_nearest_path_v2`.  Seems to be working after returning a deepcopy of the returned `Match` from `Schedule.find_match`.  [Link](https://github.com/kawa2287/title-belt-nhl/blob/15862411a6d87741a000c8b6912fc34c3cc837a8/title_belt_nhl/schedule.py#L142C17-L142C39)

File: title_belt_nhl/schedule.py (hunk 1 above)
Comment (elffaW) [reply title_belt_nhl/schedule.py]: Meh it's probably the right way to do it, just felt strange calling it with `find_nearest_path_v2([[start_match]])`.

**Review by elffaW — APPROVED** [review]

**Pull request merged**